### ISSUE TITLE
Support hover with hyphenated words by registering an `EvaluatableExpressionProvider`

### ISSUE THREAD

**Issue by GitMensch** [OlegKunitsyn/gnucobol-debug#105]
Variable view and expressions work fine with all kinds of variable names, but the hovering does not work with hyphenated words like `X-P1` (it does work if the variable is renamed to `P1`). As COBOL words very commonly are hyphenated I see this as quite an important issue to solve.

**Comment by GitMensch** [comment]
Found the issue: The vscode debug browser ui used to split words by the following:

```typescript
		// Match any character except a set of characters which often break interesting sub-expressions
		let expression: RegExp = /([^()\[\]{}<>\s+\-/%~#^;=|,`!]|\->)+/g;
```

But since vscode 1.43 there is a new [Debug hover API](https://code.visualstudio.com/updates/v1_43#_debugger-extension-api) to let language providers - or debugging extensions - have something to say about what is an actual word that is (likely to be) resolved as an expression.

I suggest to do the same as the the [MockDebug](https://github.com/microsoft/vscode-mock-debug/blob/35e0a6c2bc69b0a1464b857ad1e6cb55ada11d87/src/activateMockDebug.ts#L98) using the existing word definition of the current document for this information instead of the "the debugging extension does not tell me" fallback.

Nearly finished piece of code for extension.ts, activate() within context.subscriptions.push:

```typescript
	// override VS Code's default implementation of the debug hover by using document object's word position
	// which uses the wordpattern (defined in cobol.configuration.json)
	languages.registerEvaluatableExpressionProvider(supported_language_identifiers_here(), {
		provideEvaluatableExpression(document: vscode.TextDocument, position: vscode.Position): vscode.ProviderResult<vscode.EvaluatableExpression> {
			const wordRange = document.getWordRangeAtPosition(position);
			return wordRange ? new vscode.EvaluatableExpression(wordRange) : undefined;
		}
	});
```

A possibly more extensive option is found in https://github.com/RechInformatica/rech-cobol-debugger/blob/1497badc3347489d5930c1587d96f9ff3f1cd22d/src/provider/CobolEvaluatableExpressionProvider.ts (but I'd personally make the "is this part of a comment line" query depending on a setting, I see that some would find it confusing while other would find it fine to see variables in comments, too.

In general this extension already knows which "names" are variables, so it should be able to use the existing variable list to check for a name to pass back (after getting it from the document).

A _later_ version ideally also handles the context, for example:

```cobol
    move some-nice-word
    *> using the right index, of course:
    (some-other-word, 55)
    *> with the right part
    (12:word-len)
```

In this case hovering over `some-other-word` or `word-len` should of course only return those, but hovering over `some-nice-word` would then return `some-nice-word (some-other-word, 55) (12:word-len)` (this extension can already handle that).

**Comment by GitMensch** [comment]
Ping @OlegKunitsyn any thoughts on this?

**Comment by GitMensch** [comment]
Ping  @OlegKunitsyn, I think the possible best option for now is to register the `EvaluatableExpressionProvider` and explicit tell the debugger api about the COBOL words this extension knows about to be possible for evaluation in a given source (the list is already available in the extension, it just has to be passed on).

**Comment by GitMensch** [comment]
Friendly ping @OlegKunitsyn - especially the `EvaluatableExpressionProvider` should be useful (not sure if @marcsosduma may want to have a look at that)

**Comment by marcsosduma** [comment]
Hey! I added the code you showed and it seems to be working. It's on my
fork of Professor Oleg's branch.

File: extension.ts

[image: image.png]
Sorry! Code in portuguese.



Em dom., 9 de jul. de 2023 às 12:09, Simon Sobisch ***@***.***>
escreveu:

**Comment by GitMensch** [comment]
Can you please add a PR for that change so it is easier to compare / pull in?

Did you only add the hyphens into the `EvaluatableExpressionProvider` (possibly from the document's definition) or did you use the internal word list so something like `VAR (START-POS + 5: LEN` can be evaluated with a hover, too?

**Comment by marcsosduma** [comment]
Hi Simon!
I just added the code you indicated to EvaluatableExpressionProvider.

Em seg., 10 de jul. de 2023 às 04:26, Simon Sobisch <
***@***.***> escreveu:

**Comment by GitMensch** [comment]
Thanks for the note, please PR for easier integration.

**Comment by marcsosduma** [comment]
Hello! Oleg and Simon. I sent the PR as requested.

Em seg., 10 de jul. de 2023 às 11:10, Simon Sobisch <
***@***.***> escreveu:

**Comment by GitMensch** [comment]
done with merged #134
